server/app/models/host_node.rb

```diff
@@ -110,10 +113,58 @@ def attributes_from_docker(attrs)
   end
 
   # @return [Boolean]
+  def active?
+    self.availability == Availability::ACTIVE
+  end
+
+  # @return [Boolean]
+  def drain?
+    self.availability == Availability::DRAIN
+  end
+
+  # @return [Symbol]
+  def status
+    if self.node_id.nil?
+      return :created # node created with token, but agent has not yet connected
+    elsif !self.connected
+      return :offline # not yet connected by NodePlugger, or disconnected by NodeUnplugger
+    elsif !self.updated
+      return :connecting # connected by NodePlugger, waiting for /nodes/update RPC
+    elsif self.drain?
+      return :drain
+    else
+      return :online # connected by NodePlugger, updated by /nodes/update RPC
+    end
+  end
+
+  # @return [String]
+  def websocket_error
+    if self.connected
+      return nil
+    elsif !self.websocket_connection
+      return "Websocket is not connected"
+    elsif !self.disconnected_at || self.connected_at > self.disconnected_at
```

Comment:
Yes, particularly since the `connected_at` is from the agent clock, and the `disconnected_at` is from the server clock...

This used to differentiate between the close_code/reason set in `WebsocketBackend#on_open` -> `Agent::NodePlugger#reject!` vs `WebsocketBackend#on_close` -> `Agent::NodeUnplugger#unplug!`... I could add some new `self.websocket_connection.opened` flag instead, set to false if the open was rejected?